Support downloading subsequent newer versions of the app

hackergrrl opened the pull request:
Fixes #540.

This branch adds support for downloading multiple upgrades in a row. Previously the upgrade state machine would download at most **one** upgrade before locking itself into a state where it would download no more unless the app was restarted.

hackergrrl committed:
fix: upgrade semver sorting

hackergrrl committed:
fix: add upgrade options listener only once

hackergrrl committed:
fix: check for upgrades after current d/l finishes

hackergrrl committed:
fix: prioritize Downloading over ReadyToUpgrade

hackergrrl committed:
fix: reset upgrade state on download completion

hackergrrl committed:
fix: remove extra setState; fix param on a setState

hackergrrl committed:
fix: remove setState(IDLE) on error

gmaclennan commented on src/backend/lib/upgrade-download.js:
```diff
@@ -181,6 +183,12 @@ class Search extends EventEmitter {
   }
 
+  reset() {
+    if (this.state !== UpgradeState.Search.Searching) return;
```
My assumption would be that if the Search is in the `IDLE` state, then `reset()` would start the search (e.g. transition the Search state to `SEARCHING`), but `reset()` will only work if it is already `SEARCHING`. Is this how we want it to work? Should be documented for any future maintainers.

hackergrrl committed:
test: fixes

These are some fixes re: recent changes to the upgrade state machines
that needed correction in the tests. These fixes are pretty hacky, but I
didn't feel it made sense to invest time into rewriting them right now.

hackergrrl committed:
fix: don't show search:IDLE as an error state

This is normal behaviour: the frontend will generally do a 'get-state'
request before search has initialized, revealing it as IDLE.

gmaclennan reviewed (changes requested):
I think the original issue is because `DOWNLOADED` is not a state, it is an event. As far as I can see, we don't read or use this state anywhere. I think there are two possible states of the `Download` class:

1. `IDLE`
2. `DOWNLOADING`

The `download()` method transitions from `IDLE` --> `DOWNLOADING` and the events `error` or `download-complete` transition from `DOWNLOADING` --> `IDLE`. (I know we don't model those events as EventEmitter events, I'm referring them to abstract events caused by either the download failing or the download completing successfully).

In the future we might add a `cancelDownload()` method which would similarly transition `DOWNLOADING` --> `IDLE`.

I think the need for this reset method comes from modelling `DOWNLOADED` as a state, and what this PR does is essentially ensure that that state only exists for a single tick, because it immediately transitions that state to `IDLE`, so do we need it in the first place? It seems to add code complexity.

What does the search reset do? If the search is in the `SEARCHING` state does stopping and starting it force it to start finding things again?

hackergrrl committed:
fix: remove DOWNLOADED upgrade state

hackergrrl committed:
clean up UpgradeManager download logic

hackergrrl committed:
remove unused code

hackergrrl committed:
allow UpgradeCheck to re-check + choose latest version

hackergrrl commented:
>  I think there are two possible states of the Download class:   IDLE    DOWNLOADING

Ooh, I like this. I wrote this patch as a hack to try and fix a bug without *really* thinking through the bigger picture. I'll revise to include this idea.

> I think the need for this reset method comes from modelling DOWNLOADED as a state, and what this PR does is essentially ensure that that state only exists for a single tick, because it immediately transitions that state to IDLE, so do we need it in the first place? It seems to add code complexity.

The *actual* reason (which wasn't made obvious by just this diff) is that download options accumulate in the state's `context` object, which was confusing the logic in UpgradeManager, which decides which upgrade to download. I've revised how this works, so now the reset() logic isn't necessary.

The most recent changes also include a fix where, if there is a redownload (or multiple upgrade options in storage), the newest one wasn't being selected.

gmaclennan reviewed (approved):
Ok on a quick read-through this looks good, we can catch any issues we might have missed in QA on the share-apk-p2p branch. To address (not in this PR): put all the version compare logic in one place, to make it easier to implement https://github.com/digidem/mapeo-mobile/issues/578
